libs/pq_proto/src/lib.rs

```diff
@@ -672,6 +675,18 @@ pub struct WalSndKeepAlive {
     pub request_reply: bool,
 }
 
+#[derive(Debug)]
+pub struct InterpretedWalRecordsBody<'a> {
+    /// End of raw WAL in [`Self::data`]
+    pub streaming_lsn: u64,
+    /// Current end of WAL on the server
+    pub safekeeper_wal_end_lsn: u64,
+    /// Start LSN of the next record in PG WAL.
+    /// Is 0 if the portion of PG WAL did not contain any records.
+    pub next_record_lsn: u64,
+    pub data: &'a [u8],
```

Comment:
Was wondering what these fields were even for, since I couldn't see them used anywhere. I see that they're decoded by `rust-postgres`. Unfortunate that we have to patch external crates like that, but seems okay until we can switch the protocol.

It's worth mentioning in a struct comment though, especially since any changes here need to be mirrored there.

Reply:
* `streaming_lsn`: Used by the walreceiver connection manager when selecting safekeepers. I don't really see why we don't use the last record LSN instead, but it preserves the existing behaviour.
* `safekeeper_wal_end_lsn`: This is the commit LSN of the safekeeper. Again, it's used by the walreceiver connection manager when selecting safekeepers
* `next_record_lsn`: This is the LSN of the next record after `streaming_lsn`. We use this during pageserver ingest to advance last record LSN for cases when records were filtered out.

Reply:
Yeah, thanks -- I was initially more curious about why searching for uses of these fields didn't show anything in my IDE. It's (clearly) because the Pageserver uses the `rust-postgres` variants rather than the ones here. It might be worth mentioning on this struct that it has a twin struct in `rust-postgres` which is used on the Pageserver side, and that any changes must be made to both.

Reply:
Added a comment in https://github.com/neondatabase/neon/pull/9746/commits/6a0fb944e05b121ca8d71e60b28cc2b029291706

Reply:
>  I don't really see why we don't use the last record LSN instead

+1